Add support to ExportedProgram as input to torch.onnx.dynamo_export

thiagocrepaldi opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #111498
* __->__ #111497

Fixes #109889

This PR adds `torch.export.export` as another `FXGraphExtractor` implementation. `torch.onnx.dynamo_export` automatically uses this new FX tracer when a `torch.export.ExportedProgram` is specified as `model`

Implementation is back compatible, thus non `ExportedProgram` models are handled the exact same way as before

Commit:
Add support to ExportedProgram as input to torch.onnx.dynamo_export

Fixes #109889

[ghstack-poisoned]

Commit:
Update on "Add support to ExportedProgram as input to torch.onnx.dynamo_export"


Fixes #109889

[ghstack-poisoned]

BowenBao commented on torch/onnx/_internal/io_adapter.py:
```diff
@@ -487,3 +540,44 @@ def apply(self, model_outputs: Any) -> Sequence[Any]:
+    """Prepend model parameters and buffers to the user input.
+
+    :func:`torch.export.export` lifts model parameters and buffers as model input, thus, they
+    must be added to the user input before the model  is executed.
```
```suggestion
    must be added to the user input before the model is executed.
```
nit

BowenBao replied:
Are the prepended params and buffers part of onnx model inputs? Do people need to supply them through input_dict for ort.run call?

thiagocrepaldi replied:
Yes, they are. They are supplied through `ort.run`

As an example, 

```python
        class Model(torch.nn.Module):
            def __init__(self):
                self.normal = torch.distributions.normal.Normal(0, 1)
                super().__init__()

            def forward(self, x):
                return self.normal.sample(x.shape)

        x = torch.randn(2, 3)
        exported_program = torch.export.export(Model(), args=(x,))
        _ = torch.onnx.dynamo_export(
            exported_program,
            x,
        )
```
results in a `GraphModule`

```python
ExportedProgram:
    class GraphModule(torch.nn.Module):
        def forward(self, arg0_1: f32[], arg1_1: f32[], arg2_1: f32[2, 3]):
            # File: /home/justinchu/anaconda3/envs/onnx/lib/python3.10/site-packages/torch/distributions/normal.py:70, code: return torch.normal(self.loc.expand(shape), self.scale.expand(shape))
            expand: f32[2, 3] = torch.ops.aten.expand.default(arg0_1, [2, 3]);  arg0_1 = None
            expand_1: f32[2, 3] = torch.ops.aten.expand.default(arg1_1, [2, 3]);  arg1_1 = None
            normal: f32[2, 3] = torch.ops.aten.normal.Tensor_Tensor(expand, expand_1);  expand = expand_1 = None
            return (normal,)

Graph signature: ExportGraphSignature(parameters=[], buffers=['L__self___normal_loc', 'L__self___normal_scale'], user_inputs=['arg2_1'], user_outputs=['normal'], inputs_to_parameters={}, inputs_to_buffers={'arg0_1': 'L__self___normal_loc', 'arg1_1': 'L__self___normal_scale'}, buffers_to_mutate={}, backward_signature=None, assertion_dep_token=None)
Range constraints: {}
Equality constraints: []
```

with 1 user input and 2 buffers. When converting that as a ONN graph, we have all 3 as inputs:

```bash
[name: "arg0_1"
type {
  tensor_type {
    elem_type: 1
    shape {
    }
  }
}
, name: "arg1_1"
type {
  tensor_type {
    elem_type: 1
    shape {
    }
  }
}
, name: "arg2_1"
type {
  tensor_type {
    elem_type: 1
    shape {
      dim {
        dim_value: 2
      }
      dim {
        dim_value: 3
      }
    }
  }
}
]
```

BowenBao replied:
I think we should put those back as initializers, can be in follow up.

thiagocrepaldi replied:
That is a good point. Will do it separately. 

Adding it back makes the ONNX file complete and executable by any runtime, but then we have the issue with models >2GB. On the other hand, if the user wants to save the onnx model (and discard `ExportOutput`), then all saved buffers/initializers are lost. Maybe we have to serialize `ExportOutput` in the  ExportedProgram` as input approach

BowenBao commented on torch/onnx/_internal/fx/torch_export_graph_extractor.py:
```diff
@@ -0,0 +1,72 @@
+        return self.pre_export_passes(options, model, model.graph_module, updated_model_args)  # type: ignore[return-value]
+
+    @_beartype.beartype
+    def pre_export_passes(
```
Some passes still needed I believe, what is the plan for follow-ups?

thiagocrepaldi replied:
Any in particular comes to mind?
I am adding them as I discover the need for them.  Right now there is only one simple test, so maybe some passes are missed

My idea was to create a PR in which we can reuse all `TestFxToOnnxWithOnnxRuntime` for testing with both `fx_tracer=dynamo_export_graph_extractor.DynamoExport` and `fx_tracer=torch_export_graph_extractor.TorchExport`. That would flush everything we need to support

BowenBao replied:
type promotion is immediate one I can think of. Yep we can play catch-up with test coverage and bench results.

BowenBao commented on torch/onnx/_internal/io_adapter.py:
```diff
@@ -487,3 +540,44 @@ def apply(self, model_outputs: Any) -> Sequence[Any]:
+        self,
+        model_args: Sequence[Any],
+        model_kwargs: Mapping[str, Any],
+        model: Optional[
```
pass model in `__init__` to keep `apply` signature unchanged?

thiagocrepaldi replied:
I was wondering the same. I think this is a good idea.

BowenBao commented on torch/onnx/_internal/exporter.py:
```diff
@@ -350,16 +351,31 @@ class ResolvedExportOptions(ExportOptions):
+            torch_export_graph_extractor,
+        )
+
+        assert options is not None, "ExportOptions must be specified."
```
why this change?

thiagocrepaldi replied:
We were initializing `ExportOptions` in different points of the code. I found out this when writing a unit test that would detect an invalid `fx_tracer` and `model` combination. With this change, `def dynamo_export` is responsible to initialize `options` in a single spot

BowenBao replied:
The original code looks clean too as all the None check are postponed to this point inside `ResolvedExportOptions`.

I don't have a strong opinion. If you are continuing with this change, please update argument `options: Optional[Union[ExportOptions, "ResolvedExportOptions"]] = None,` to `options: Union[ExportOptions, "ResolvedExportOptions"]`. Doesn't make sense to allow None that will immediately trigger assertion error.

BowenBao reviewed (commented):
A very quick skim through and some comments.

Commit:
Update on "Add support to ExportedProgram as input to torch.onnx.dynamo_export"


Fixes #109889

This PR adds `torch.export.export` as another `FXGraphExtractor` implementation. `torch.onnx.dynamo_export` automatically uses this new FX tracer when a `torch.export.ExportedProgram` is specified as `model`

Implementation is back compatible, thus non `ExportedProgram` models are handled the exact same way as before

[ghstack-poisoned]

thiagocrepaldi commented on torch/onnx/_internal/exporter.py:
```diff
@@ -350,16 +351,31 @@ class ResolvedExportOptions(ExportOptions):
-        self, options: Optional[Union[ExportOptions, "ResolvedExportOptions"]]
+        self,
+        options: Optional[Union[ExportOptions, "ResolvedExportOptions"]] = None,
+        model: Optional[Union[torch.nn.Module, Callable, torch_export.ExportedProgram]] = None,  # type: ignore[name-defined]
```
This I would like to pick your brain. Adding `model` solely for validation purposes. Should I move the validation outside `ResolvedExportOptions` and let the API simpler?

Commit:
Update on "Add support to ExportedProgram as input to torch.onnx.dynamo_export"


Fixes #109889

This PR adds `torch.export.export` as another `FXGraphExtractor` implementation. `torch.onnx.dynamo_export` automatically uses this new FX tracer when a `torch.export.ExportedProgram` is specified as `model`

Implementation is back compatible, thus non `ExportedProgram` models are handled the exact same way as before

[ghstack-poisoned]

thiagocrepaldi commented:
@pytorchbot merge

huydhn commented:
@pytorchbot revert -m 'Sorry for reverting your change, it is failing ONNX test in trunk https://hud.pytorch.org/pytorch/pytorch/commit/4f42edfb6e5b703eec2a14b8933090646702c5a2, possibly a landrace' -c landrace

Commit:
Update on "Add support to ExportedProgram as input to torch.onnx.dynamo_export"


Fixes #109889

This PR adds `torch.export.export` as another `FXGraphExtractor` implementation. `torch.onnx.dynamo_export` automatically uses this new FX tracer when a `torch.export.ExportedProgram` is specified as `model`

Implementation is back compatible, thus non `ExportedProgram` models are handled the exact same way as before

[ghstack-poisoned]

thiagocrepaldi commented:
@pytorchbot merge
